nickel/src/lib.rs

```diff
@@ -0,0 +1,637 @@
+        let rt = program.eval_full_for_export().map_err(|error| Error {
+            error: Box::new(error),
+            files: program.files(),
+        })?;
+
+        Ok(Expr { rt })
+    }
+
+    /// Evaluate a Nickel program to weak head normal form (WHNF).
+    ///
+    /// The result of this evaluation is a null, bool, number, string,
+    /// enum, record, or array. In case it's a record, array, or enum
+    /// variant, the payload (record values, array elements, or enum
+    /// payloads) will be left unevaluated.
+    ///
+    /// Together with the expression, this returns a Nickel virtual machine that
+    /// can be used to further evaluate unevaluated sub-expressions.
+    pub fn eval_shallow(&mut self, src: &str) -> Result<(VirtualMachine, Expr), Error> {
+        let mut program: Program<CacheImpl> = Program::new_from_source(
+            Cursor::new(src),
+            self.name.as_deref().unwrap_or("<source>"),
+            self.trace.clone(),
+            NullReporter {},
+        )?;
+        program.add_import_paths(self.import_paths.iter());
+        let rt = program.eval().map_err(|error| Error {
+            error: Box::new(error),
+            files: program.files(),
+        })?;
+
+        Ok((VirtualMachine { program }, Expr { rt }))
+    }
+}
+
+/// A Nickel virtual machine.
+///
+/// This can be used to further evaluate unevaluated subexpressions (thunks).
+/// You can create one of these with [`Context::eval_shallow`].
+pub struct VirtualMachine {
+    program: Program<CacheImpl>,
+}
+
+impl VirtualMachine {
+    /// Evaluate an expression to weak head normal form (WHNF).
+    ///
+    /// This has an effect only if the expression is a thunk (see
+    /// [`Expr::is_thunk`]).
+    ///
+    /// The result of this evaluation is a null, bool, number, string,
+    /// enum, record, or array. In case it's a record, array, or enum
+    /// variant, the payload (record values, array elements, or enum
+    /// payloads) will be left unevaluated.
+    pub fn eval_shallow(&mut self, expr: Expr) -> Result<Expr, Error> {
+        // The term should already be closurized because there's no
+        // public API for making an `Expr` with an unclosurized term.
+        // But Program::eval_closure wants an explicit Closure.
+        let clos = Closure {
+            body: expr.rt,
+            env: Environment::new(),
+        };
+
+        let rt = self.program.eval_closure(clos).map_err(|error| Error {
+            error: Box::new(error.into()),
+            files: self.program.files(),
+        })?;
+
+        Ok(Expr { rt })
+    }
+}
+
+/// A Nickel expression.
+///
+/// This might be fully evaluated (for example, if you got it from [`Context::eval_deep`])
+/// but might have unevaluated thunks (if you got it from [`Context::eval_shallow`]).
+#[derive(Clone)]
+pub struct Expr {
+    rt: RichTerm,
+}
+
+/// A Nickel record.
+///
+/// Fields can be accessed either by name or by index.
+pub struct Record<'a> {
```

Comment:
What are those lifetimes `'a`? Since the internal fields are not visible in the documentation, we should probably explain roughly what they represent.

Reply:
Yup, looks good